lib/modules/firestore/Transaction.js

```diff
@@ -0,0 +1,179 @@
+/**
+ * @flow
+ * Firestore Transaction representation wrapper
+ */
+import { mergeFieldPathData } from './utils';
+import { buildNativeMap } from './utils/serialize';
+
+import type Firestore from './';
+import type { TransactionMeta } from './TransactionHandler';
+import type DocumentReference from './DocumentReference';
+import type DocumentSnapshot from './DocumentSnapshot';
+import { isObject, isString } from '../../utils';
+import FieldPath from './FieldPath';
+import { getNativeModule } from '../../utils/native';
+
+type Command = {
+  type: 'set' | 'update' | 'delete',
+  path: string,
+  data: ?{ [string]: any },
+  options: ?{ merge: boolean },
+};
+
+type SetOptions = {
+  merge: boolean,
+};
+
+/**
+ * @class Transaction
+ */
+export default class Transaction {
+  _pendingResult: ?any;
+  _firestore: Firestore;
+  _meta: TransactionMeta;
+  _commandBuffer: Array<Command>;
+
+  constructor(firestore: Firestore, meta: TransactionMeta) {
+    this._meta = meta;
+    this._commandBuffer = [];
+    this._firestore = firestore;
+    this._pendingResult = undefined;
+  }
+
+  /**
+   * -------------
+   * INTERNAL API
+   * -------------
+   */
+
+  /**
+   * Clears the command buffer and any pending result in prep for
+   * the next transaction iteration attempt.
+   *
+   * @private
+   */
+  _prepare() {
+    this._commandBuffer = [];
+    this._pendingResult = undefined;
+  }
+
+  /**
+   * -------------
+   *  PUBLIC API
+   * -------------
+   */
+
+  /**
+   * Reads the document referenced by the provided DocumentReference.
+   *
+   * @param documentRef DocumentReference A reference to the document to be retrieved. Value must not be null.
+   *
+   * @returns Promise<DocumentSnapshot>
+   */
+  get(documentRef: DocumentReference): Promise<DocumentSnapshot> {
+    // todo validate doc ref
+    return getNativeModule(this._firestore).transactionGetDocument(
+      this._meta.id,
+      documentRef.path
+    );
+  }
+
+  /**
+   * Writes to the document referred to by the provided DocumentReference.
+   * If the document does not exist yet, it will be created. If you pass options,
+   * the provided data can be merged into the existing document.
+   *
+   * @param documentRef DocumentReference A reference to the document to be created. Value must not be null.
+   * @param data Object An object of the fields and values for the document.
+   * @param options SetOptions An object to configure the set behavior.
+   *        Pass {merge: true} to only replace the values specified in the data argument.
+   *        Fields omitted will remain untouched.
+   *
+   * @returns {Transaction}
+   */
+  set(
+    documentRef: DocumentReference,
+    data: Object,
+    options?: SetOptions
```

Comment:
We need to add SetOptions to the main `DocumentReference.set()` method as well as this was added to the Firestore SDKs recently.

Reply:
Scrap that, was thinking of something else